### ISSUE TITLE
mutate.Time should be lazy

### ISSUE THREAD

**Issue by imjasonh** [google/go-containerregistry#1168]
[`pkg/v1/mutate.Time`](https://pkg.go.dev/github.com/google/go-containerregistry/pkg/v1/mutate#Time) sets all timestamps in an image to the given time, typically to make the image more reproducible. For timestamps in layers, it calls [`layerTime`](https://github.com/google/go-containerregistry/blob/v0.6.0/pkg/v1/mutate/mutate.go#L376) which reads the layer contents and writes them to a buffered `tar.Writer`. This means that if the layer is huge, we may run out of memory buffering its contents.

We should consider making this a lazy transformation, so that layer mutations aren't made until they're read. We did a similar thing in [`pkg/v1/cache`](https://github.com/google/go-containerregistry/blob/v0.6.0/pkg/v1/cache/cache.go#L70), where layer contents aren't cached until they're read via `Compressed` or `Uncompressed`.

**Comment by jonjohnsonjr** [comment]
At one point I experimented with this. The API was gross due to streaming layers, but that was so long ago there might be a less naive way to do it: https://github.com/jonjohnsonjr/go-containerregistry/commit/542d74bc764fc0f023888bdd7cad46b79841ce82

**Comment by imjasonh** [comment]
Yeah I like this `Mutation` concept. It doesn't help that much with `mutate.Time` though since that wants to deal with files in the tar archive, and that's still a huge pain to extract from the `v1.Image`, especially streamily.

Maybe we could expose some adapter that takes a `func(tar.Header) (tar.Header, error)`?

**Comment by jonjohnsonjr** [comment]
For the layerTime, I ended up having a `ReadCloser` -> `ReadCloser`: https://github.com/jonjohnsonjr/go-containerregistry/commit/542d74bc764fc0f023888bdd7cad46b79841ce82#diff-05a233d3122c90ba0cf547c623e3a34194cf56c4ff05cdf414d5a265300b31e6R39

**Comment by bh-tt** [comment]
I've improved this partially in https://github.com/GoogleContainerTools/kaniko/pull/3347, but I wonder if I should submit the PR here?
